go/internal/otel/main.go

```diff
@@ -0,0 +1,141 @@
+package otel
+
+import (
+	"context"
+	"encoding/hex"
+	"fmt"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	otelCode "go.opentelemetry.io/otel/codes"
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace"
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
+	"go.opentelemetry.io/otel/sdk/resource"
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+	semconv "go.opentelemetry.io/otel/semconv/v1.4.0"
+	"go.opentelemetry.io/otel/trace"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/status"
+)
+
+var tracer trace.Tracer
+
+func decodeTraceID(encodedSpanID string) (t trace.TraceID, err error) {
+	var spanBytes []byte
+	spanBytes, err = hex.DecodeString(encodedSpanID)
+	if err != nil {
+		err = fmt.Errorf("failed to decode spanID: %w", err)
+		return
+	}
+	copy(t[:], spanBytes)
+	return
+}
+
+func decodeSpanID(encodedSpanID string) (s trace.SpanID, err error) {
+	var spanBytes []byte
+	spanBytes, err = hex.DecodeString(encodedSpanID)
+	if err != nil {
+		err = fmt.Errorf("failed to decode spanID: %w", err)
+		return
+	}
+	copy(s[:], spanBytes)
+	return
+}
+
+// ServerGrpcInterceptor is a gRPC server interceptor for tracing and optional metadata-based context enhancement.
+func ServerGrpcInterceptor(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
```

Comment:
do we need to modify interceptors in other services to make this work the same way?

Reply:
yes, pr in golang/python is already done